### PR TITLE
Add tags for elastic constitutive relations

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

<s>Adds a simple tag that holds an `Elasticity::ConstitutiveRelations::ConstitutiveRelation` as a `std::unique_ptr`</s> (since it's an abstract base class, similar to our equations of state). Also adds a tag that constructs from options, e.g. from the analytic solution.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Comment by nilsvu** [comment]
@geoffrey4444 @moxcodes could you do the review for this? As always, feel free to forward to another reviewer if you don't have the time.

File: src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp
Comment (moxcodes) [inline src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp]: this is an initialization tag, right? It might be helpful to include that classification in the dox.

File: tests/Unit/PointwiseFunctions/Elasticity/ConstitutiveRelations/Test_Tags.cpp
Comment (moxcodes) [inline tests/Unit/PointwiseFunctions/Elasticity/ConstitutiveRelations/Test_Tags.cpp]: I don't think there's any way for these to get reordered in this test, but I think it's good testing philosophy generally to pass in different numerical arguments so that it makes sure that the values ended up in the parameters you intended.

**Review by moxcodes — COMMENTED** [review]
looks good, only nitpicks.

File: src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp
Comment (nilsvu) [reply src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp]: This is my understanding of our classification: Parallel components and actions can have `initialization_tags` which will be inserted into the initial DataBox, and which will be removed by the `RemoveOptionsAndTerminatePhase` action. They must be option-creatable, because there's no other way to insert them into the DataBox at that point. But also the `const_global_cache_tags` are created from options, and that's where this constitutive relation tag is intended to live (but doesn't have to). So I think the classification is that all `initialization_tags` and `const_global_cache_tags` must be "option-creatable". I actually intend to add an `OptionCreatableTag` protocol (once #1769 is merged) to make this requirement clear.
I think the documentation here makes clear that the tag is option-creatable, but let me know if you agree with my understanding.

**Comment by nilsvu** [comment]
@geoffrey4444 please take a look at this PR

File: src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp
Comment (moxcodes) [reply src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp]: Ah, understood. Yes, this is clearly option-creatable, so that is fine.

**Review by moxcodes — DISMISSED** [review]
go ahead and squash

**Review by moxcodes — DISMISSED** [review]

**Comment by kidder** [comment]
@geoffrey4444 please look at

**Comment by nilsvu** [comment]
@geoffrey4444 please take a look

**Comment by nilsvu** [comment]
@geoffrey4444 ping

**Comment by nilsdeppe** [comment]
Would @tomwlodarczyk be a good person to provide a second review?

**Comment by nilsvu** [comment]
Yes @tomwlodarczyk please take a look and comment on anything that’s not entirely clear.

File: src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp
Comment (tomwlodarczyk) [inline src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp:39]: What kind of metavariables have you had in mind here? Also, I take it the create_from_options here is the one used below with Parallel::create_from_options<>(), although I don't get how? I have found a Parallelization.md file that I hope will clear things up for me.

**Review by tomwlodarczyk — DISMISSED** [review]

File: src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp
Comment (nilsvu) [reply src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp:39]: The `Metavariables` template parameter is part of the compile-time interface that the `Parallel::create_from_options` function expects. Our [dev guide on the parallelization infrastructure](https://spectre-code.org/dev_guide_parallelization_foundations.html) is indeed a good place to start :)

**Comment by nilsvu** [comment]
@sxs-collaboration/spectre-core-devs This PR has been reviewed by @moxcodes and @tomwlodarczyk. Could you take another look? Cheers!

**Comment by nilsvu** [comment]
@sxs-collaboration/spectre-core-devs please take a look. (rebased again to make sure the CI runs on current develop)

**Review by moxcodes — DISMISSED** [review]

File: src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp
Comment (kidder) [reply src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp:39]: since #2026 has been merged, delete the template parameter and add `static constexpr bool pass_metavariables = false;` as we've made it optional to know the metavariables.

Comment (kidder) [reply src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp:39]: also why is this a separate tag? why can't you just put `create_from_options` into the previous tag?  if you want them as separate tags, then this one should be the `SimpleTag` and the previous one should be the `BaseTag`

Comment (nilsvu) [reply src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp:39]: - It needs to be a separate tag because it retrieves the constitutive relation from an option-created provider (e.g. the analytic solution) that it is templated on. The `Tags::ConstitutiveRelation` simple tag does not need to know where the constitutive relation it holds comes from; in particular it shouldn't need to be templated on the provider every time it's used to retrieve the constitutive relation from the DataBox
- The `Tags::ConstitutiveRelation` should be a simple tag because it know what type it holds (but is agnostic to where its value came from). It can be used stand-alone to put a constitutive relation in the DataBox if it is not being retrieved from a provider.

This reflects my understanding of DataBox tags, please correct me if I got something wrong

Comment (kidder) [reply src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp:39]: did you check that you can retrieve from the DataBox via `ConstitutiveRelation` if it is put into the box with `ConstitutiveRelationFrom`?

Comment (nilsvu) [reply src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp:39]: Yes! I added a test in a fixup commit. However, this is just the standard behavior of our DataBox, so it's probably not necessary to test that for every tag.

Comment (kidder) [reply src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp:39]: It seems to me `ConstitutiveRelation` in Elasticity is similar to `EquationOfState` in GRMHD.  You either specify it freely , or you need to grab it from the analytic solution.  If this is the case, please do what is done in the GRMHD executable for an analytic solution

**Comment by nilsdeppe** [comment]
@nilsleiffischer can you rebase this on develop? It has a conflict

**Comment by nilsdeppe** [comment]
@moxcodes @tomwlodarczyk @kidder  this got rebased and needs your eyes again :)

**Review by moxcodes — DISMISSED** [review]

File: src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp
Comment (nilsvu) [reply src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp:39]: Not sure if you still have objections to this @kidder. Here's why I chose the current design:
- We need to retrieve the constitutive relation in the system fluxes, which use an `argument_tags`-style databox interface. Therefore the argument-tag can't be a type-erased base tag with a strongly typed simple tag in the databox without templating the system on the constitutive relation type; at least current that's not supported because the flux must build an `apply` function but can't get the type information from the base tag without the full tags list.
- We could template the elasticity system on the constitutive relation as is done in GRMHD. Or we store the constitutive relation in a `std::unique_ptr` to its base class. I chose the latter because we only need the base class interface when using it. Is there an advantage to the former? Why did you choose that design for GRMHD?

Comment (nilsvu) [reply src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp:39]: @kidder I made a mistake earlier. Turns out there is a way to avoid templating the system on the constitutive relation but keep the constitutive relation tag strongly-typed. This design is almost the same as the EOS tags, with the difference that the base tag is typed with the base class `ConstitutiveRelation<Dim>`. This allows us to use the base class in function signatures where the concrete subclass is irrelevant, but it's still possible to retrieve the strongly-typed constitutive relation where the full DataBox tags list is available. The changes reverted a lot of the previous commit's diff, so I pushed a rewritten commit instead of a fixup. Please take a look!

**Comment by nilsvu** [comment]
@kidder could you take a look at this please?

**Review by moxcodes — DISMISSED** [review]

File: src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp
Comment (kidder) [reply src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp:39]: @nilsleiffischer I think you need to choose whether the DataBox holds a `std::unique_ptr<ConstitutiveRelation<Dim>>` or a concrete type derived from `ConstitutiveRelation<Dim>`
If you choose the pointer, then I would rename `Tags::ConstitutiveRelationBase` to `Tags::ConstitutiveRelation` and give it the `name` function you want.  If you want to create it directly from the option file, then add an option tag and add the appropriate `create_from_options` (see e.g. `Tags::StepController` from `Time/Tags.hpp`), and put it in the `initialization_tags` of the initialization action; if you want to create it from an analytic solution, the initialization action just grabs it from the solution and puts it into the DataBox.     If you want to have a concrete type, then keep the  `ConstitutiveRelationBase` as a `db::BaseTag` and always retrieve by the base tag (either for `get` or `argument_tags`).  Only the initialization action will need to know the concrete type and you should be able to do this the same way as `EquationOfState` is added in `Evolution/Initialization/ConservativeSystem.hpp`

Comment (nilsvu) [reply src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp:39]: @kidder To help this discussion I opened #2100 that is based on this PR. Let's talk on Monday.

Comment (nilsvu) [reply src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp:39]: This is what I do in #2100:
- The `Tags::ConstitutiveRelationFrom` is placed in the ConstGlobalCache, so it's available through the databox and it's created from the analytic solution through `create_from_options`.
- It can be retrieved either through `Tags::ConstitutiveRelation<ConcreteSubclass>` or `Tags::ConstitutiveRelationBase<Dim>`. With the former you get the object of the concrete subclass type, and with the latter you get it of the virtual base class type. So when the subclass type is not available, you still know it's a `ConstitutiveRelation` and can call the base class functions on it.
- I use the base tag in the elliptic fluxes computer, which only needs to know the object is a `ConstitutiveRelation` but doesn't care which subclass.

**Comment by nilsvu** [comment]
@kidder since #2104 got stuck I implemented the other option we discussed, i.e. store the constitutive relation in a `unique_ptr`. We can adjust this later, e.g. if we manage to get base tags working for it. But for now I'd like to have a working version so we can add an elasticity executable. Please take a look :)

**Comment by nilsdeppe** [comment]
Maybe you could provide me with the code example that's giving you trouble here so I can better understand what's going on?

**Review by kidder — DISMISSED** [review]

**Comment by nilsvu** [comment]
@nilsdeppe and I discussed this in a call and opened #2111 and #2112. Once those two PRs are merged we can use base tags in this PR.

**Review by kidder — DISMISSED** [review]

**Comment by nilsvu** [comment]
@moxcodes and @tomwlodarczyk I know you have approved this uncountably many times before, but could you take another (hopefully final) look at this? We have it figured out now I think :)

File: src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp
Comment (nilsdeppe) [inline src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp]: you're still having a simple tag inherit from a simple tag

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp
Comment (nilsvu) [reply src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp]: do you have another suggestion? There will be different ways of getting the `Tags::ConstitutiveRelation` into the DataBox:
- Put this `ConstitutiveRelationFrom` tag into the ConstGlobalCache so the constitutive relation is retrieved from the analytic solution or analytic data
- Load the constitutive relation directly from the input file

Comment (nilsdeppe) [reply src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp]: Why not specify the constitutive relation in the metavariables like we do for equations of state?

**Comment by nilsvu** [comment]
- [x] Opened #2170 as another dependency of this PR. For such a tiny diff this PR has generated a surprising amount of controversy 😄

**Comment by nilsvu** [comment]
This is ready for review again since #2170 was merged. @nilsdeppe @kidder could you take a look since you requested the recent changes?

File: src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp
Comment (nilsdeppe) [inline src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp]: `noexcept`

**Review by nilsdeppe — DISMISSED** [review]
One minor suggestion (that you may squash), but this is nice. I think it's a clean way to handle this :)

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**
